src/relay/qnn/op/convolution.cc

```diff
@@ -50,12 +50,15 @@ bool QnnConv2DRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
   if (data == nullptr || weight == nullptr) return false;
   const auto* param = attrs.as<Conv2DAttrs>();
   ICHECK(param != nullptr) << "Conv2DAttrs cannot be nullptr.";
-  ICHECK(data->dtype == DataType::Int(8) || data->dtype == DataType::UInt(8))
-      << "Expected qnn conv2d type(int8, uint8) for input but was " << data->dtype;
-  ICHECK(weight->dtype == DataType::Int(8) || weight->dtype == DataType::UInt(8))
-      << "Expected qnn conv2d type(int8, uint8) for weight but was " << weight->dtype;
-  ICHECK(param->out_dtype == DataType::Int(16) || param->out_dtype == DataType::Int(32))
-      << "Expected qnn conv2d type(int32, int16) for output but was " << param->out_dtype;
+  ICHECK(data->dtype == DataType::Int(8) || data->dtype == DataType::UInt(8) ||
+         data->dtype == DataType::Int(16))
+      << "Expected qnn conv2d type(int8, uint8, int16) for input but was " << data->dtype;
```

Comment:
Yeah that would work, I think! I eyeballed the conv2d file a bit more and I think in most of the cases where there is casting into int32, the weights would need to be _very_ big for it to overflow int32, I'm worried about https://github.com/apache/tvm/blob/main/src/relay/qnn/op/convolution.cc#L566-L571 though, I feel like multiplying two zero points that are probably 16 bit values and then multiplying through with the shape of the weights could overflow so maybe worth casting this into int64 to be on the safe side?

Reply:
To keep the current behaviour I limited in t a minimum of 32 bits for the existing cases, and for cases that we expect the output to be more than 32 (for example in TFLite when we adjust outputs to be 64 bits), that it will grow accordingly.

Can you have another look and see whether it is looks ok now?